### ISSUE TITLE
StackOverflowError running ch03-recommender

### ISSUE THREAD

**Issue by karth295** [sryza/aas#113]
I'm running Spark-on-YARN with Spark 2.2.0 and Hadoop 2.8.1. I ran the job using: `spark-submit ch03-recommender/target/ch03-recommender-2.0.0-jar-with-dependencies.jar`. I'm using pretty small VMs -- 2 VMs with 2 cores and 8gb of memory each.

I've run this job several times, and seen a few executors die each time because of the stackoverflow. Apparently tasks succeed on retries -- the job completes (with some logspam).

Here's the full stacktrace in the stderr of the executors: https://pastebin.com/Hf4Ueip9.

Feel free to close this issue if it's something weird about my environment.

**Comment by karth295** [comment]
I think I got to the bottom of this. ALS's iterative algorithm creates a lot of RDD lineage, and overtime this leads to a stackoverflow: https://stackoverflow.com/questions/31484460/spark-gives-a-stackoverflowerror-when-training-using-als.

I must have only seen this on the smaller VM type because the JVM must be configured with smaller thread stacks.

ALS does indeed [set a default checkpoint interval](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala#L862), but you need to enable checkpointing in the SparkContext.

**Comment by karth295** [comment]
tl;dr: this issue is that you need to set a checkpoint dir in this example. Since you're already assuming `/user/ds` exists, a subdirectory of that is probably reasonable.

**Comment by srowen** [comment]
Yes you can shorten the checkpoint interval. Or you can allow the JVM to use larger thread stack sizes with -Xss. That does mean using somewhat more memory.

**Comment by karth295** [comment]
Well a checkpoint interval of 10 is reasonable but you need to set a checkpoint directory in here: https://github.com/sryza/aas/blob/master/ch03-recommender/src/main/scala/com/cloudera/datascience/recommender/RunRecommender.scala#L20.

I think this issue is still valid and needs fixing.

**Comment by srowen** [comment]
OK, that's fine. I'll add a line to set the checkpoint dir for this example. I ran it successfully with and without a checkpoint dir on Spark 2.2.
